### PR TITLE
Join order optimizer asan bug Follow up

### PR DESCRIPTION
**Pull request by Tmonster**
Follow up to https://github.com/duckdb/duckdb/pull/11719

This PR removes the JoinNode class completely. In addition, it removes a lot of other dead code that should have been removed a while ago. This solution passes a const reference to the dynamic programming table of plans to the query_graph_manager so that plan reconstruction can directly reference the dynamic programming table.  The reconstruction code can then lookup the left and right JoinRelationSets by looking into the DP table.

### CODE REVIEW

File: src/include/duckdb/optimizer/join_order/plan_enumerator.hpp [hunk 1]
```diff
@@ -58,13 +59,12 @@ class PlanEnumerator {
...
 	//! the newly created plan, or an existing plan)
-	unique_ptr<JoinNode> EmitPair(JoinRelationSet &left, JoinRelationSet &right,
-	                              const vector<reference<NeighborInfo>> &info);
+	DPJoinNode &EmitPair(JoinRelationSet &left, JoinRelationSet &right, const vector<reference<NeighborInfo>> &info);
 	//! Tries to emit a potential join candidate pair. Returns false if too many pairs have already been emitted,
 	//! cancelling the dynamic programming step.
 	bool TryEmitPair(JoinRelationSet &left, JoinRelationSet &right, const vector<reference<NeighborInfo>> &info);
 
-	unique_ptr<JoinNode> CreateJoinNodeFromDPJoinNode(DPJoinNode dp_node);
+	//	unique_ptr<JoinNode> CreateJoinNodeFromDPJoinNode(DPJoinNode dp_node);
```
Comment (lnkuiper) [inline src/include/duckdb/optimizer/join_order/plan_enumerator.hpp:67]: Should this comment be removed?

File: src/optimizer/join_order/plan_enumerator.cpp [hunk 2]
```diff
@@ -91,22 +91,9 @@ void PlanEnumerator::GenerateCrossProducts() {
...
-		D_ASSERT(left_dp_join_node->second);
-		D_ASSERT(right_dp_join_node->second);
-		auto left = CreateJoinNodeFromDPJoinNode(*left_dp_join_node->second);
-		auto right = CreateJoinNodeFromDPJoinNode(*right_dp_join_node->second);
-		auto res = make_uniq<JoinNode>(dp_node.set, dp_node.info, std::move(left), std::move(right), dp_node.cost);
-		res->cardinality = dp_node.cardinality;
-		return res;
-	}
+const reference_map_t<JoinRelationSet, unique_ptr<DPJoinNode>> &PlanEnumerator::GetPlans() const {
+	//	optional_ptr<reference_map_t<JoinRelationSet, unique_ptr<DPJoinNode>>> res = &plans;
```
Comment (lnkuiper) [inline src/optimizer/join_order/plan_enumerator.cpp:95]: Should this comment be removed?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Looks great after the refactor! I just have two remarks, and then this is ready to go:

**Review by lnkuiper — APPROVED** [review]
LGTM now, thanks!

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**